### PR TITLE
Stabilise vec::remove_item

### LINKED ISSUE #40062: Tracking issue for Vec::remove_item
Implemented in https://github.com/rust-lang/rust/pull/40325

```rust
impl<T: PartialEq> Vec<T> {
    pub fn remove_item(&mut self, item: &T) -> Option<T> { /*…*/ }
}

### PR DESCRIPTION
**Pull request by Dylan-DPC-zz**
Closes #40062

r? @alexcrichton

### CODE REVIEW

**Comment by Dylan-DPC** [comment]
uggh some git mess up.. let me fix it :D

File: src/liballoc/tests/vec.rs [hunk 1]
```diff
@@ -131,6 +131,22 @@ fn test_extend_ref() {
     assert_eq!(v, [1, 2, 3, 4, 5, 6, 7]);
 }
 
+#[test]
+fn test_remove_item() {
+    let mut y = vec![1,2,3];
+    v.remove_item(&1);
+
+    assert_eq!(v.len(2));
```
Comment (jfrimmel) [inline src/liballoc/tests/vec.rs]:
```suggestion
    assert_eq!(v.len(), 2);
```

File: src/liballoc/tests/vec.rs [hunk 2]
```diff
@@ -131,6 +131,22 @@ fn test_extend_ref() {
...
+    assert_eq!(v, [1,3]);
```
Comment (jfrimmel) [inline src/liballoc/tests/vec.rs]: Shouldn't this be `assert_eq!(v, [2, 3])`?

File: src/liballoc/tests/vec.rs [hunk 3]
```diff
@@ -131,6 +131,22 @@ fn test_extend_ref() {
     assert_eq!(v, [1, 2, 3, 4, 5, 6, 7]);
 }
 
+#[test]
+fn test_remove_item() {
+    let mut y = vec![1,2,3];
```
Comment (jfrimmel) [inline src/liballoc/tests/vec.rs]:
```suggestion
    let mut v = vec![1,2,3];
```

File: src/liballoc/tests/vec.rs [hunk 4]
```diff
@@ -131,6 +131,22 @@ fn test_extend_ref() {
...
+    v.remove_item(&1);
+
+    assert_eq!(v.len(2));
+    assert_eq!(v, [1,3]);
+
+    let mut w = vec![1,2,3];
+    w.remove_item(&4);
+
+    assert_eq!(v.len(3));
```
Comment (jfrimmel) [inline src/liballoc/tests/vec.rs]:
```suggestion
    assert_eq!(w.len(), 3);
```

**Review by jfrimmel — CHANGES_REQUESTED** [review]
I'm pretty sure, that the new test doesn't work in the current form :smile:

**Comment by Dylan-DPC** [comment]
@jfrimmel ye you right :D had a gcc issue so just went with the tests _in blind form_ waiting for the CI to tell me what's wrong :P fixed everything now. Thanks for the review

**Comment by Dylan-DPC** [comment]
spurious... 

@bors retry

**Comment by timvermeulen** [comment]
Many people suggested in the tracking issue to have this method take a `&U` with `T: PartialEq<U>` instead of a `&T`, shouldn't this be implemented before we stabilize this? In particular, this doesn't compile at the moment:
```rust
let mut vec = vec![String::from("foo")];
vec.remove_item("foo");
//              ^^^^^ expected struct `std::string::String`, found `str`
```

**Comment by Dylan-DPC** [comment]
Created a new PR that made that change. Blocked on #67873

**Comment by alexcrichton** [comment]
Looks like FCP has finished on https://github.com/rust-lang/rust/issues/40062 and while there's been discussion since it doesn't seem ground-shaking, so I'm going to approve. 

@bors: r+

**Comment by alexcrichton** [comment]
@bors: r-

(CI failures)

**Comment by Dylan-DPC** [comment]
ye forgot to remove 1 feature gate.. pushing it now :)

**Comment by kornelski** [comment]
Please don't stabilise it without fixing the name!

Please don't call it just `remove_item`, as this is the same name as in other languages that _can_ remove item from an array by presenting an item from the same array. Rust's version can't do that due to borrow checker and it's a trap for users.

**Pull request merged**